src/ca/idcert.rs

```diff
@@ -0,0 +1,668 @@
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the certificate is a correctly issued certificate.
+    ///
+    /// Note this check is used to check that an EE certificate in an RFC8183,
+    /// or RFC6492 message is validly signed by the TA certificate that was
+    /// exchanged.
+    ///
+    /// This check assumes for now that we are always dealing with V3
+    /// certificates and AKI and SKI have to match.
+    fn validate_issued(
+        &self, issuer: &IdCert
+    ) -> Result<(), ValidationError> {
+        // Authority Key Identifier. Must be present and match the
+        // subject key ID of `issuer`.
+        if let Some(aki) = self.authority_key_id {
+            if aki != issuer.subject_key_id {
+                return Err(ValidationError);
+            }
+        } else {
+            return Err(ValidationError);
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the certificate is a valid CA certificate.
+    ///
+    /// Checks the parts that are common in normal and trust anchor CA
+    /// certificates.
+    fn validate_ca_basics(&self) -> Result<(), ValidationError> {
+        // 4.8.1. Basic Constraints: For a CA it must be present (RFC6487)
+        // und the “cA” flag must be set (RFC5280).
+        if let Some(ca) = self.basic_ca {
+            if ca {
+                return Ok(());
+            }
+        }
+
+        Err(ValidationError)
+    }
+
+    /// Validates the certificate’s signature.
+    fn validate_signature(
+        &self, issuer: &IdCert
+    ) -> Result<(), ValidationError> {
+        self.signed_data.verify_signature(issuer.subject_public_key_info())
+    }
+}
+
+
+//--- Deref, AsRef
+
+impl ops::Deref for IdCert {
+    type Target = TbsIdCert;
+
+    fn deref(&self) -> &Self::Target {
+        &self.tbs
+    }
+}
+
+impl AsRef<IdCert> for IdCert {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl AsRef<TbsIdCert> for IdCert {
+    fn as_ref(&self) -> &TbsIdCert {
+        &self.tbs
+    }
+}
+
+
+//--- Deserialize and Serialize
+
+#[cfg(feature = "serde")]
+impl serde::Serialize for IdCert {
+    fn serialize<S: serde::Serializer>(
+        &self, serializer: S
+    ) -> Result<S::Ok, S::Error> {
+        let bytes = self.to_captured().into_bytes();
+        let str = base64::encode(&bytes);
+        str.serialize(serializer)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for IdCert {
+    fn deserialize<D: serde::Deserializer<'de>>(
+        deserializer: D
+    ) -> Result<Self, D::Error> {
+        use serde::de;
+
+        let some = String::deserialize(deserializer)?;
+        let dec = base64::decode(&some).map_err(de::Error::custom)?;
+        let b = Bytes::from(dec);
+        IdCert::decode(b).map_err(de::Error::custom)
+    }
+}
+
+//--- PartialEq and Eq
+
+impl PartialEq for IdCert {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_captured().into_bytes().eq(&other.to_captured().into_bytes())
```

Comment:
The implementation via `SignedData::data` means that two IdCerts are equal even if they are signed with different signatures. This may or may not be what you want semantically – given that there is a separate `TbsIdCert` type, I think it probably isn’t. If you only want to compare the content, maybe do it on `TbsIdCert`?

(If you want to add `PartialEq` and `Eq` impls on `SignedData` that would be fine by me.)